File: internal/manifest/guideline.go
package manifest

import "strings"

// GuidelineFor returns the guideline/usage doc associated with a component, if
// one exists. It matches docs whose id begins with the component's id-derived
// prefix and looks like a guideline page (e.g.
// "components-alert-guidelines-usage--docs" for component "components-alert").
func GuidelineFor(docs *Docs, c Component) (Doc, bool) {
	if docs == nil {
		return Doc{}, false
	}
	// Component ids look like "components-buttons-mainbutton"; guideline doc ids
	// look like "components-<name>-guidelines-usage--docs". Try a direct prefix
	// match first, then a looser name-based match.
	prefix := c.ID + "-guidelines"
	for _, d := range docs.Docs {
		if strings.HasPrefix(d.ID, prefix) {
			return d, true
		}
	}

	name := strings.ToLower(c.Name)
	for _, d := range docs.Docs {
		id := strings.ToLower(d.ID)
		if strings.Contains(id, "guidelines") && strings.Contains(id, name) {
			return d, true
		}
	}
	return Doc{}, false
}

File: src/manifest/guideline.ts
import type { Component, Doc, Docs } from "./types.js";

/**
 * Returns the guideline/usage doc associated with a component, if one exists.
 * It matches docs whose id begins with the component's id-derived prefix and
 * looks like a guideline page (e.g. "components-alert-guidelines-usage--docs"
 * for component "components-alert").
 */
export function guidelineFor(docs: Docs | undefined, c: Component): Doc | undefined {
  const all = docs?.docs;
  if (!all) return undefined;

  // Component ids look like "components-buttons-mainbutton"; guideline doc ids
  // look like "components-<name>-guidelines-usage--docs". Try a direct prefix
  // match first, then a looser name-based match.
  const prefix = `${c.id ?? ""}-guidelines`;
  if (c.id) {
    for (const d of Object.values(all)) {
      if (d.id?.startsWith(prefix)) return d;
    }
  }

  const name = (c.name ?? "").toLowerCase();
  if (name) {
    for (const d of Object.values(all)) {
      const id = (d.id ?? "").toLowerCase();
      if (id.includes("guidelines") && id.includes(name)) return d;
    }
  }
  return undefined;
}
